content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc

```diff
@@ -0,0 +1,424 @@
+which in turn means those precious CPU cores can spend more cycles processing your actual business logic.
+
+But also for an actual application itself it can be beneficial to keep an eye -- and systematically track -- object allocations,
+as regressions there lead to increased GC pressure in turn to higher latencies and reduced through-put.
+
+The key for tracking object allocations with JFR are the `jdk.ObjectAllocationInNewTLAB` and `jdk.ObjectAllocationOutsideTLAB` events,
+which are emitted when
+
+* Object allocation triggered the creation of a new thread-local allocation buffer (TLAB) 
+* An object got allocated outside of the threads TLAB
+
+[NOTE]
+.Thread-local allocation buffers (TLAB)
+====
+When creating new object instances on the heap,
+this primarily happens via _thread-local allocation buffers_.
+A TLAB is a pre-allocated memory block that's exclusively used by a single thread.
+Creating new objects within a TLAB can happen without costly synchronization with other threads.
+Once a thread's current TLAB capacity is about to be exceeded with a new object allocation,
+a new TLAB will be allocated for that thread.
+In addition, large objects can also be directly allocated, outside of the TLAB.
+
+To learn more about TLAB allocation, refer to  https://shipilev.net/jvm/anatomy-quarks/4-tlab-allocation/[part #4] of Aleksey Shipilёv's "JVM Anatomy Quark" blog series.
+====
+
+Note these events don't allow for tracking of each individual object allocation,
+as multiple objects will be allocated within a TLAB before a new one is required and thus the `jdk.ObjectAllocationInNewTLAB` event will be emitted.
+But as that event exposes the size of the new TLAB, we can keep track of the overall amount of memory that's allocated.
+
+In that sense, `jdk.ObjectAllocationInNewTLAB` represents a sampling of object allocations,
+which means we need to collect a reasonable number of events to identify those locations in the program which are the sources of high memory allocation and thus trigger new TLAB creation often.
+
+So let's start and work on a test that'll spot regressions in terms of object allocations of one of the todo app's API methods, `GET /todo/{id}`.
+To identify a baseline of the allocation to be expected,
+we first invoke that method in a loop and print out the actual allocation values.
+This should happen in intervals, e.g. every 10,000 invocations,
+so to average out numbers from individual calls.
+
+[source, java]
+----
+@Test
+@EnableEvent("jdk.ObjectAllocationInNewTLAB") // <1>
+@EnableEvent("jdk.ObjectAllocationOutsideTLAB")
+public void retrieveTodoBaseline() throws Exception {
+  Random r = new Random();
+
+  HttpClient client = HttpClient.newBuilder()
+      .build();
+
+  for (int i = 1; i<= 100_000; i++) {
+    executeRequest(r.nextInt(20) + 1, client);
+
+    if (i % 10_000 == 0) {
+      jfrEvents.awaitEvents(); // <2>
+
+      long sum = jfrEvents.filter(this::isObjectAllocationEvent)  // <3>
```

Comment:
I would find one call to .filter(...) and combined expression more natural here? WDYT?

Reply:
Hum, hum. Couldn't use the method refs then, I think?